src/main/java/com/epam/aidial/core/storage/BlobStorage.java

```diff
@@ -0,0 +1,199 @@
+import org.jclouds.ContextBuilder;
+import org.jclouds.blobstore.BlobStore;
+import org.jclouds.blobstore.BlobStoreContext;
+import org.jclouds.blobstore.domain.Blob;
+import org.jclouds.blobstore.domain.BlobMetadata;
+import org.jclouds.blobstore.domain.MultipartPart;
+import org.jclouds.blobstore.domain.MultipartUpload;
+import org.jclouds.blobstore.domain.PageSet;
+import org.jclouds.blobstore.domain.StorageMetadata;
+import org.jclouds.blobstore.domain.StorageType;
+import org.jclouds.blobstore.domain.Tier;
+import org.jclouds.blobstore.domain.internal.BlobMetadataImpl;
+import org.jclouds.blobstore.options.ListContainerOptions;
+import org.jclouds.blobstore.options.PutOptions;
+import org.jclouds.io.ContentMetadata;
+import org.jclouds.io.ContentMetadataBuilder;
+import org.jclouds.io.payloads.BaseMutableContentMetadata;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+@Slf4j
+public class BlobStorage {
+
+    private final BlobStoreContext storeContext;
+    private final BlobStore blobStore;
+    private final String bucketName;
+
+    public BlobStorage(Storage config) {
+        ContextBuilder builder = ContextBuilder.newBuilder(config.getProvider());
+        if (config.getEndpoint() != null) {
+            builder.endpoint(config.getEndpoint());
+        }
+        builder.credentials(config.getIdentity(), config.getCredential());
+        this.storeContext = builder.buildView(BlobStoreContext.class);
+        this.blobStore = storeContext.getBlobStore();
+        this.bucketName = config.getBucket();
+        createBucketIfNeeded(config);
+    }
+
+    /**
+     * Initialize s3 multipart upload
+     *
+     * @param fileName    name of the file, for example: data.csv
+     * @param path        absolute path according to the bucket, for example: Users/user1/files/input data
+     * @param contentType MIME type of the content, for example: text/csv
+     */
+    public MultipartUpload initMultipartUpload(String fileName, String path, String contentType) {
+        BlobMetadata metadata = buildBlobMetadata(fileName, path, contentType, bucketName);
+        return blobStore.initiateMultipartUpload(bucketName, metadata, PutOptions.NONE);
+    }
+
+    /**
+     * Upload part/chunk of the file
+     *
+     * @param multipart MultipartUpload that chunk related to
+     * @param part      chunk number, starting from 1
+     * @param buffer    data
+     */
+    public MultipartPart storeMultipartPart(MultipartUpload multipart, int part, Buffer buffer) {
+        return blobStore.uploadMultipartPart(multipart, part, new BufferPayload(buffer));
+    }
+
+    /**
+     * Commit multipart upload.
+     * This method must be called after all parts/chunks uploaded
+     */
+    public void completeMultipartUpload(MultipartUpload multipart, List<MultipartPart> parts) {
+        String etag = blobStore.completeMultipartUpload(multipart, parts);
+        log.info("Stored etag: " + etag);
```

Comment:
I guess the log message has a little sense. Please remove the line or return it in the method if we support concurrent updates.

Reply:
fixed